service/storage_service.cc

```diff
@@ -3011,14 +3016,65 @@ future<> storage_service::join_cluster(sharded<db::system_distributed_keyspace>&
+                    "The recovery_leader option will be ignored. If you are trying to run the Raft recovery procedure, "
+                    "ensure you delete raft_group_id and truncate system.discovery.",
+                    _db.local().get_config().recovery_leader(), _group0->load_my_id());
+        } else {
+            recovery_leader_id = locator::host_id(utils::UUID(_db.local().get_config().recovery_leader()));
+            if (recovery_leader_id != my_host_id() && !loaded_endpoints.contains(recovery_leader_id)) {
+                throw std::runtime_error(format("Recovery leader {} unrecognised as a cluster member", recovery_leader_id));
+            }
+            recovery_leader_ip = recovery_leader_id == my_host_id() ?
+                    get_broadcast_address() : loaded_endpoints[recovery_leader_id].endpoint;
+            initial_contact_nodes = std::unordered_set{recovery_leader_ip};
+
+            if (_group0->client().in_recovery()) {
+                throw std::runtime_error(format(
+                        "Entered RECOVERY mode and set recovery_leader to {}. RECOVERY mode is used in the old "
+                        "recovery procedure, while recovery_leader is used in the new procedure. If raft-based "
+                        "topology is enabled in the whole cluster, use the new procedure. Otherwise, use the old one.",
+                        recovery_leader_id));
+            }
+            if (!_sys_ks.local().bootstrap_complete()) {
+                throw std::runtime_error("Cannot bootstrap in Raft recovery");
+            }
+            if (!co_await _sys_ks.local().load_topology_features_state()) {
+                throw std::runtime_error("Cannot start in Raft recovery - raft-based topology has not been enabled");
+            }
+            if (_db.local().get_config().force_gossip_topology_changes()) {
+                throw std::runtime_error("Cannot force gossip topology changes in Raft recovery");
+            }
+        }
+    }
+
     auto loaded_peer_features = co_await _sys_ks.local().load_peer_features();
     slogger.info("initial_contact_nodes={}, loaded_endpoints={}, loaded_peer_features={}",
             initial_contact_nodes, loaded_endpoints | std::views::keys, loaded_peer_features.size());
     for (auto& x : loaded_peer_features) {
         slogger.info("peer={}, supported_features={}", x.first, x.second);
     }
 
-    if (_group0->client().in_recovery()) {
+    if (recovery_leader_id) {
+        // The new Raft recovery procedure.
+        slogger.info("Raft recovery with recovery leader {}/{}", recovery_leader_id, recovery_leader_ip);
+        auto g0_info = co_await _group0->discover_group0(std::vector{recovery_leader_ip}, _qp);
+        if (g0_info.id.uuid() != recovery_leader_id.uuid()) {
+            throw std::runtime_error(fmt::format(
+                    "Raft recovery - found group 0 {} with leader {}/{} not matching recovery leader {}/{}.",
```

Comment:
Perhaps more details how to get out of this if it happens --- I imagine it could happen if someone doesn't follow the procedure completely, in particular they don't restart the recovery leader first or don't wait for it to become fully restarted before moving on to the next node (e.g. if they concurrently restart everything; some other node could become the leader of discovery)

in any case if they mess up, they should probably start again? But this time ensure correct order of restarting nodes

Reply:
> Perhaps more details how to get out of this if it happens

I'll add more details.

> I imagine it could happen if someone doesn't follow the procedure completely, in particular they don't restart the recovery leader first or don't wait for it to become fully restarted before moving on to the next node (e.g. if they concurrently restart everything; some other node could become the leader of discovery)

I'll check what happens in these scenarios.

> in any case if they mess up, they should probably start again? But this time ensure correct order of restarting nodes

Yes, they should start again.